src/EventStore.Core/TransactionLog/Chunks/ChunkFooter.cs

```diff
@@ -50,38 +56,67 @@ public readonly long
 			MapCount = mapSize / posMapSize;
 		}
 
-		public byte[] AsByteArray() {
-			var array = new byte[Size];
-			using (var memStream = new MemoryStream(array))
-			using (var writer = new BinaryWriter(memStream)) {
-				var flags = (byte)((IsCompleted ? 1 : 0) | (IsMap12Bytes ? 2 : 0));
-				writer.Write(flags);
-				writer.Write(PhysicalDataSize);
-				if (IsMap12Bytes)
-					writer.Write(LogicalDataSize);
-				else
-					writer.Write((int)LogicalDataSize);
-				writer.Write(MapSize);
-
-				memStream.Position = Size - ChecksumSize;
-				writer.Write(MD5Hash);
+		public ChunkFooter(ReadOnlySpan<byte> source) {
+			Debug.Assert(source.Length >= Size);
+
+			SpanReader<byte> reader = new(source);
+			byte flags = reader.Read();
+
+			IsCompleted = (flags & 1) is not 0;
+			IsMap12Bytes = (flags & 2) is not 0;
+			PhysicalDataSize = reader.ReadLittleEndian<int>();
+			LogicalDataSize = IsMap12Bytes
+				? reader.ReadLittleEndian<long>()
+				: reader.ReadLittleEndian<int>();
+
+			MapSize = reader.ReadLittleEndian<int>();
+			reader.ConsumedCount = Size - ChecksumSize;
+			MD5Hash = reader.ReadToEnd().ToArray();
+
+			var posMapSize = IsMap12Bytes ? PosMap.FullSize : PosMap.DeprecatedSize;
+			if (MapSize % posMapSize is not 0) {
+				throw new Exception(string.Format("Wrong MapSize {0} -- not divisible by PosMap.Size {1}.", MapSize,
+					posMapSize));
 			}
 
+			MapCount = MapSize / posMapSize;
+		}
+
+		static int IBinaryFormattable<ChunkFooter>.Size => Size;
+
+		public void Format(Span<byte> destination) {
```

Comment:
this writes less data than `Size` bytes, should we fill the remainder of Size with 0s? or alternatively write Size zeroes to the span first? same for header

Reply:
I though about that. Checksum doesn't include these bits. They considered as reserved so no matter how they filled.

Reply:
fair enough, i think it just means if we start using some of those bits later we will tell the footer some version number so it knows whether it should look at them or not, but perhaps we ought to do that in any case